internal/referral/rewarder.go

```diff
@@ -0,0 +1,121 @@
+// Package referral ...
+package referral
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/Decentr-net/vulcan/internal/blockchain"
+	"github.com/Decentr-net/vulcan/internal/blockchain/rest"
+	"github.com/Decentr-net/vulcan/internal/storage"
+)
+
+// Rewarder ...
+type Rewarder struct {
+	storage storage.Storage
+	bmc     blockchain.Blockchain
+	brc     *rest.BlockchainRESTClient
+
+	senderReward   int
+	receiverReward int
+	uPDVThreshold  int
+}
+
+// NewRewarder creates a new instance of Rewarder.
+func NewRewarder(s storage.Storage, b blockchain.Blockchain, brc *rest.BlockchainRESTClient,
+	senderReward, receiverReward, uPDVThreshold int) *Rewarder {
+	return &Rewarder{
+		storage:        s,
+		bmc:            b,
+		brc:            brc,
+		senderReward:   senderReward,
+		receiverReward: receiverReward,
+		uPDVThreshold:  uPDVThreshold,
+	}
+}
+
+// Run runs the rewarder check referral status loop.
+func (r *Rewarder) Run(ctx context.Context, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	go func(ticker *time.Ticker) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ticker.C:
+				r.do(ctx)
+			}
+		}
+	}(ticker)
```

Comment:
shouldn't we run `do`  at start as well?

Reply:
Added